### ISSUE TITLE
set-psreadlineoption typo in -Colors example and wrong? default value on -HistoryNoDuplicates

### ISSUE THREAD

**Issue by UberKluger** [MicrosoftDocs/PowerShell-Docs#6219]
The example for -Colors (PSH 6+) has "e[91m", should be "`e[91m" (missing back-tick).

The default value for -HistoryNoDuplicates is listed as False but invoking Set-PSReadLineOption without -HistoryNoDuplicates does not reset the option while using -HistoryNoDuplicates:$False does reset it (value apparently $False in both cases). Would it not be better to describe the default value of switch parameters as Null if they have no effect when not included and/or behave differently when explicitly assigned $False?

(Begin: a general comment regarding switch parameters in all documentation.)
I note that some cmdlet switch parameters have "default" values which are defined externally. For example, -Confirm defaults to a value determined by $ConfirmPreference. The documentation for Remove-Item lists the default value as False. It does include links to other relevant documents (including descriptions of preference variables) but has no direct mention that the "default" value can be altered by the things described in this other documentation (let alone specifying exactly which thing affects it).

While I understand that cluttering up documentation by repeating basic behaviour that is succinctly described elsewhere can be annoying to readers, it should be remembered that documentation has 2 basic purposes, reference and education. Those readers who would find verbosity annoying are probably experienced users just confirming the use of a parameter (reference), in which case they will likely skip directly to the required information. However, for less experienced users (reading for education), omitting relevant material on the assumption that "everybody already knows it" can lead to confusion and frustration when it doesn't work as anticipated. This is particularly true when they already have something that isn't working and are trying to determine the cause of the problem.
(End: general comment)

So, with regard to both reference and education, it might be useful in this case (and others) to include in the switch parameter definition (and not as an example somewhere else in the document) a note about how to reset a persistent value (by using "-HistoryNoDuplicates:$False").

**Comment by sdwheeler** [comment]
I will fix the escape character issue. That is a typo related to markdown rendering.

The convention in our docs is that the default value of a parameter is the value that is used by the cmdlet when the parameter is not provided. By definition, switch parameters are boolean values. When you do not include a switch parameter the value is `false`. When you include a switch parameter the value is `true`. There are cases where an example would helpful, especially where the negative logic can be confusing. If you see a need like that you can file a specific issue or a PR to add the example.

PSReadline is problematic because it is using switch parameters to change global settings, not to select a different behavior of the cmdlet. What we need are examples that show how to change those settings.

The `-Confirm` parameter is a bit of a special case, since behavior is affected by `$ConfirmPreference` and the impact level defined by the cmdlet. Again, examples are probably the best way to illustrate the scenarios.

**Comment by UberKluger** [comment]
I think you have misinterpreted my description.

I am aware that switch parameters are boolean and that the default value is
$False. My point was that some switch parameters appear to be ternary in
that they can have 3 values (for 3 different behaviours), being

$True for parameter present (with or without explicit $True)
$False for parameter absent
$OtherFalse for parameter present with explicit $False (and behaving
differently than parameter absent $False).

I listed -HistoryNoDuplicates as an example though, upon further reading,
my assertion that the "value" of -Confirm was externally defined is not
entirely accurate. -Confirm is like -HistoryNoDuplicates in that it can
have 3 values (as described above), being

$True = always confirm (as if $ConfirmPreference were "Low")
$False = use $ConfirmPreference to determine confirmation based upon risk
level (as defined by the cmdlet)
$OtherFalse = never confirm, irrespective of risk level (as if
$ConfirmPreference were "None")

Other examples are:
Parameter        $True                                            $False
                $OtherFalse
-Debug           use "Inquire"* instead of $DebugPreference       use
$DebugPreference      use "SilentlyContinue" instead of $DebugPreference
-Verbose         use "Continue"* instead of $VerbosePreference    use
$VerbosePreference    use "SilentlyContinue" instead of $VerbosePreference
-WhatIf          use $True instead of $WhatIfPreference           use
$WhatIfPreference     use $False instead of $WhatIfPreference

*Error in "about_Preference_Variables" [
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7].
The example implies that -Debug overrides $DebugPreference with "Continue"
instead of "Inquire" as described in the "valid values" section (and as
tested in PS 5.1.17134.858).
 Omission in "about_CommonParameters" [
https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_commonparameters?view=powershell-7].
The description of -Verbose does not detail how it overrides
$VerbosePreference (by using the value "Continue" as determined by testing
in PS 5.1.17134.858).
(I list these issues here because I found them while researching this
reply, if they *NEED* to be brought up as an "issue" then I can do that.)

These examples are just the ones I found by a quick search through
about_CommonParameters (where this ternary nature is actually described).
There are probably many others that I wouldn't find until I used them in an
unusual way (like turning OFF -HistoryNoDuplicates) and which therefore
would remain incorrectly (incompletely?) described. The question that I
would ask now is:

If these are truly Boolean parameters, how can the cmdlet tell the
difference between $False and $OtherFalse?

Maybe all switch parameters are actually Boolean? (nullable boolean) or at
least behave that way when being parsed from the command line (e.g. an
integer with 3 values -1,0,1). As such my suggestion of using Null as the
value of an absent switch parameter would be quite appropriate.

Whether or not it is determined that altering the basic definition of a
switch parameter (perhaps to include/add the notion of a nullable/ternary
switch parameter but then possibly requiring ALL documentation to be
updated) is worth the effort (because only a few cmdlets are significantly
affected), it would still be useful to include a note within the individual
cmdlet documentation if the effect IS significant (at the point where the
parameter is described like it is in about_CommonParameters). The changes
could be implemented progressively by having a policy of fixing them only
when the documentation is being changed for some other reason, such as for
new cmdlets (get it fully described the first time), new cmdlet features
(even if not directly affecting the parameters in question) or editorial
fixups/raised issue (functional errors, spelling, grammar, typos or even
this issue raised in the future). All documentation could be flagged as
potentially needing revision and then revised/cleared the next time it
becomes appropriate to update it for any reason. It wouldn't be too
difficult to check whether any switch parameters had ternary behaviour at
that time.

I believe that all documentation should be complete and not rely on any
"assumed knowledge", at least insofar as the thing being documented is
concerned. So while it might be reasonable to assume that someone reading
Powershell documentation has a basic (or even somewhat advanced)
understanding of computers, if a cmdlet parameter can produce 3 different
results depending upon how it is specified (or omitted) then all 3
possibilities should be described in the relevant part of the documentation
irrespective of the "intended" audience, even if one of the
possibilities is "has no effect" (typically for "omitted"). Though
including "no effect" descriptions might seem excessive, it clearly defines
what is supposed to happen when the parameter is omitted (i.e. nothing),
just like the "This page intentionally left blank" notations.

An example (not of "no effect") would be Format-Table where the behaviour
in the absence of a property parameter (not a switch) is well defined but
varies substantially depending upon the type of object being processed (a
fact which is somewhat glossed over in the documentation), while including
a property parameter actually produces a more stable output (possibly with
empty fields if the processed objects don't have the all relevant
properties). Thus, a clear description (located in the -Property parameter
description) of exactly what to expect in the absence of a property
parameter would be appropriate. Something like "either uses the
formatting/properties as defined by the specified view parameter or by the
default view specifications located in .... if no view parameter is given
or, if no view specification applies, attempts to display all the
properties of the first object (subject to screen width limitations, c.f.
-Wrap) and (only) the properties from subsequent objects with those names".
The current documentation doesn't mention that Format-Table will use a
-View, if specified, when describing the effect of omitting -Property, just
that -Property and -View are mutually exclusive (in a separate paragraph).
You might say "if you specify a view then that's obviously what it will
use" but my point is that documentation shouldn't rely on end users knowing
what is "obvious", it should all be explicitly stated. Skipping over the
"obvious" should be reserved for quick start guides.

Sorry, that last bit turned into a general rant about the things I don't
like with a lot of documentation but it still applies.

If replying directly is inappropriate then I shall respond within the
relevant raised issue in future. (and limit the general rants)
